src/main/kotlin/com/amazon/opendistroforelasticsearch/indexstatemanagement/util/ManagedIndexUtils.kt

```diff
@@ -338,27 +366,56 @@ fun ManagedIndexMetaData.getUpdatedManagedIndexMetaData(
 
     val updatedStateMetaData = state.getUpdatedStateMetaData(this)
     val updatedActionMetaData = action.getUpdatedActionMetaData(this, state)
+    val updatedStepMetaData = step.getStartingStepMetaData()
+
+    return this.copy(
+        stateMetaData = updatedStateMetaData,
+        actionMetaData = updatedActionMetaData,
+        stepMetaData = updatedStepMetaData
+    )
+}
+
+@Suppress("ReturnCount")
+fun ManagedIndexMetaData.getCompletedManagedIndexMetaData(
+    state: State,
+    action: Action,
+    step: Step,
+    actionMetaData: ActionMetaData
+): ManagedIndexMetaData {
     val updatedStepMetaData = step.getUpdatedManagedIndexMetaData(this)
 
-    // TODO Update Action retry logic.
+    val updatedActionMetaData = if (updatedStepMetaData.stepMetaData?.stepStatus == Step.StepStatus.FAILED) {
+        when {
+            action.config.configRetry == null -> actionMetaData.copy(failed = true)
+            actionMetaData.consumedRetries == 0 -> actionMetaData.copy(
```

Comment:
Can consumed retries be negative for special reasons (ex. ```-1```)?  If not, we should be able to cover this case
with our ```else``` check below

Reply:
I guess it shouldn't be negative but in case we do, I updated to that we set it to 1.

Reply:
Misunderstood the comment. The comment was about how we shouldn't need to check for specific consumed retries 0 condition. I removed this condition check.